### PR TITLE
Add Donation modal

### PR DESCRIPTION
**Pull request by dem4ron**
Screenshot 2023-09-21 at 11 17 21

### CODE REVIEW

File: app/css/ui-kit/buttons.css [hunk 1]
```diff
@@ -231,4 +231,10 @@
             @apply text-15;
         }
     }
+
+    .btn-casual {
+        @apply text-textColor1 font-semibold text-18;
+        @apply border-1 border-textColor1 bg-white rounded-8;
+        @apply shadow-xsZ1v3;
+    }
```
Comment (iHiD) [inline app/css/ui-kit/buttons.css]: We shouldn't include font-sizes, rounding, etc in button classes like this. The text-18 and rounded-8 etc should come from adding the `btn-l` to the element. I also think in this case, the actual button is `btn-default` or `btn-enhanced`. I think the only difference is the shadow?

File: app/css/ui-kit/filters.css [hunk 2]
```diff
@@ -15,4 +15,9 @@
     .filter-lightGold {
         filter: var(--lightGold-filter);
     }
+
+    .filter-shadow-huge {
+        -webkit-filter: drop-shadow(0px 4px 128px rgba(79, 114, 205, 0.8));
+        filter: drop-shadow(0px 4px 128px rgba(79, 114, 205, 0.8));
```
Comment (iHiD) [inline app/css/ui-kit/filters.css]: You can do this via tailwind, which removes the need for the extra webkit variant: https://tailwindcss.com/docs/drop-shadow#arbitrary-values

File: app/javascript/components/common/GraphicalIcon.tsx [hunk 3]
```diff
@@ -5,8 +5,8 @@ export function GraphicalIcon({
   icon,
   className = '',
   category,
-  width = null,
-  height = null,
+  width = 0,
+  height = 0,
```
Comment (iHiD) [inline app/javascript/components/common/GraphicalIcon.tsx]: This means something very different I think. It'll set the images to have 0px width/height if the thing isn't specified, rather than using the default of the image.

**Comment by dem4ron** [comment]
@iHiD about modal a11y: 
It seems to be a bad idea to have the same ids multiple times, but these elements get replaced, and there is only one unique id present in the DOM, and I don't really know how it could be easily bypassed in [this structure](https://github.com/exercism/website/blob/2a40119d05184754a1a6e378f2a83c80238992a6/app/javascript/components/modals/student/FinishMentorDiscussionModal.tsx#L137-L146). What do you think about this? 
I think I'll summon @SleeplessByte :)

File: app/css/ui-kit/buttons.css (hunk 1 above)
Comment (dem4ron) [reply app/css/ui-kit/buttons.css]: There are a couple differences between any of those and this one in the design. But `btn-enhanced` is the closest to this.

> We shouldn't include font-sizes, rounding, etc in button classes like this. The text-18 and rounded-8 etc should come from adding the btn-l to the element. 

Ok!

File: app/javascript/components/common/GraphicalIcon.tsx (hunk 3 above)
Comment (dem4ron) [reply app/javascript/components/common/GraphicalIcon.tsx]: You are right, took a proper look at it, and the correct default value is `undefined` if you want to get the default sizes. I updated the type signature on [GraphicalIcon.](https://github.com/exercism/website/blob/40e6f245bd3e35fa9c29d8e75034246936b1bf81/app/javascript/components/common/GraphicalIcon.tsx#L4-L10)

**Comment by SleeplessByte** [comment]
- If the question is: should you ever have two or more elements with the same ID, the answer is *no*, which I think you also say (unique).
- If the question is: is it okay to _remove_ an element with an ID to later add an element with that same ID, even if the added element isn't the same as the removed one: that's fine. I think this is the case given your comment about there only being one instance per ID in the DOM at the same time.

File: app/javascript/components/modals/student/finish-mentor-discussion-modal/DonationStep.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+//               "subscription": null
+//           }
+//       }
+//   },
+//   "userSignedIn": true,
+//   "links": {
+//       "settings": "http://local.exercism.io:3020/settings/donations",
+//       "donate": "http://local.exercism.io:3020/donate"
+//   }
+// }
```
Comment (ErikSchierboom) [inline app/javascript/components/modals/student/finish-mentor-discussion-modal/DonationStep.tsx]: I think this should be removed

**Comment by iHiD** [comment]
@dem4ron @ErikSchierboom Can we get this green pls (might just need a rebase). Then can one of you update the OP so I know what I need to do get it over the line. Thanks.

**Comment by ErikSchierboom** [comment]
My latest commit should have fixed the failing test.
@dem4ron do you know what the status is on the client-side? I think the modal was not yet shown, was it?

File: app/css/ui-kit/filters.css (hunk 2 above)
Comment (iHiD) [reply app/css/ui-kit/filters.css]: @dem4ron

File: app/helpers/react_components/student/mentoring_session.rb [hunk 5]
```diff
@@ -6,26 +6,48 @@ class MentoringSession < ReactComponent
...
+                }
+              },
+              # TODO: add correct endpoint and initial data
+              user_signed_in: user_signed_in?,
+              captcha_required: !current_user || current_user.captcha_required?,
+              recaptcha_site_key: ENV.fetch('RECAPTCHA_SITE_KEY', Exercism.secrets.recaptcha_site_key),
+              links: {
+                settings: Exercism::Routes.donations_settings_url,
+                donate: Exercism::Routes.donate_url
+              }
```
Comment (iHiD) [inline app/helpers/react_components/student/mentoring_session.rb]: This nesting is wrong. This shouldn't all be in `links`. It should be part of the main hash.

**Comment by iHiD** [comment]
@dem4ron This just needs the "don't close when clicking the background" thing setting and then I think your bit is done.

**Comment by iHiD** [comment]
And also just check my comments.

File: test/system/flows/student/mentor_finished_discussion_test.rb [hunk 6]
```diff
@@ -48,9 +48,9 @@ class MentorFinishedDiscussionTest < ApplicationSystemTestCase
           click_on "It was good!"
           fill_in "Leave #{discussion.mentor.handle} a testimonial (optional)", with: "Good mentor!"
           click_on "Finish"
-          click_on "Back to the exercise"
+          click_on "Continue to exercise"
 
-          assert_text "Nice, it looks like you’re done here!"
+          assert_text "Want to try another mentor?"
```
Comment (iHiD) [inline test/system/flows/student/mentor_finished_discussion_test.rb]: @dem4ron Why has this changed. If someone selects "It was good!", I don't think we should be taking them down the "request another mentor" path

File: app/helpers/react_components/student/mentoring_session.rb [hunk 7]
```diff
@@ -6,26 +6,44 @@ class MentoringSession < ReactComponent
...
+          exercise: SerializeMentorSessionExercise.(exercise),
+          iterations:,
+          mentor: mentor_data,
+          track_objectives: user_track&.objectives.to_s,
+          out_of_date: solution.out_of_date?,
+          videos:,
+          donation: {
+            user_signed_in: user_signed_in?,
+            captcha_required: !current_user || current_user.captcha_required?,
+            recaptcha_site_key: ENV.fetch('RECAPTCHA_SITE_KEY', Exercism.secrets.recaptcha_site_key),
```
Comment (iHiD) [inline app/helpers/react_components/student/mentoring_session.rb]:
```suggestion
            recaptcha_site_key: Exercism.secrets.recaptcha_site_key,
```

File: app/helpers/react_components/student/mentoring_session.rb [hunk 8]
```diff
@@ -6,26 +6,44 @@ class MentoringSession < ReactComponent
...
+          track: SerializeMentorSessionTrack.(track),
+          exercise: SerializeMentorSessionExercise.(exercise),
+          iterations:,
+          mentor: mentor_data,
+          track_objectives: user_track&.objectives.to_s,
+          out_of_date: solution.out_of_date?,
+          videos:,
+          donation: {
+            user_signed_in: user_signed_in?,
+            captcha_required: !current_user || current_user.captcha_required?,
```
Comment (iHiD) [inline app/helpers/react_components/student/mentoring_session.rb]:
```suggestion
            captcha_required: false
```

File: app/helpers/react_components/student/mentoring_session.rb [hunk 9]
```diff
@@ -6,26 +6,44 @@ class MentoringSession < ReactComponent
...
+          discussion: discussion ? SerializeMentorDiscussionForStudent.(discussion) : nil,
+          track: SerializeMentorSessionTrack.(track),
+          exercise: SerializeMentorSessionExercise.(exercise),
+          iterations:,
+          mentor: mentor_data,
+          track_objectives: user_track&.objectives.to_s,
+          out_of_date: solution.out_of_date?,
+          videos:,
+          donation: {
+            user_signed_in: user_signed_in?,
```
Comment (iHiD) [inline app/helpers/react_components/student/mentoring_session.rb]:
```suggestion
            user_signed_in: true,
```

File: app/images/graphics/wizard-hat.svg
Comment (iHiD) [inline app/images/graphics/wizard-hat.svg:1]: Can we remove this now?

File: app/helpers/react_components/student/mentoring_session.rb [hunk 10]
```diff
@@ -6,26 +6,44 @@ class MentoringSession < ReactComponent
...
+          exercise: SerializeMentorSessionExercise.(exercise),
+          iterations:,
+          mentor: mentor_data,
+          track_objectives: user_track&.objectives.to_s,
+          out_of_date: solution.out_of_date?,
+          videos:,
+          donation: {
+            user_signed_in: true,
+            captcha_required: false,
+            recaptcha_site_key: Exercism.secrets.recaptcha_site_key,
```
Comment (iHiD) [inline app/helpers/react_components/student/mentoring_session.rb]: Do we need this, @dem4ron as `captcha_required` is always false?

Comment (dem4ron) [reply app/helpers/react_components/student/mentoring_session.rb]: then probably we don't

File: app/images/graphics/wizard-hat.svg
Comment (dem4ron) [reply app/images/graphics/wizard-hat.svg:1]: yes, it isn't used elsewhere.

File: app/helpers/react_components/student/mentoring_session.rb (hunk 10 above)
Comment (iHiD) [reply app/helpers/react_components/student/mentoring_session.rb]: OK. I'll leave you to remove it! :)

File: .github/labels.yml [hunk 11]
```diff
@@ -175,7 +175,7 @@
 - name: "bot/bugsnag"
   description: ""
   color: "A32E92"
-  
+
```
Comment (ErikSchierboom) [inline .github/labels.yml]: Why was this changed? This file is auto-generated so shouldn't be edited directly.

File: CODE_OF_CONDUCT.md [hunk 12]
```diff
@@ -90,4 +90,4 @@ This policy was initially adopted from the Front-end London Slack community and
 A version history can be seen on [GitHub](https://github.com/exercism/website-copy/edit/main/pages/code_of_conduct.md).
 
 _This policy is a "living" document, and subject to refinement and expansion in the future.
-This policy applies to the Exercism website, the Exercism GitHub organization, any other Exercism-related communication channels (e.g.  Slack, Twitter, email) and any other Exercism entity or event._
+This policy applies to the Exercism website, the Exercism GitHub organization, any other Exercism-related communication channels (e.g. Slack, Twitter, email) and any other Exercism entity or event._
```
Comment (ErikSchierboom) [inline CODE_OF_CONDUCT.md]: This file is auto-synced from https://github.com/exercism/org-wide-files/blob/main/global-files/CODE_OF_CONDUCT.md, so any local changes will get overwritten at some point. Maybe revert?

File: app/helpers/react_components/donations/footer_form.rb [hunk 13]
```diff
@@ -17,7 +17,7 @@ def to_s
             recaptcha_site_key: ENV.fetch('RECAPTCHA_SITE_KEY', Exercism.secrets.recaptcha_site_key),
             links: {
               settings: Exercism::Routes.donations_settings_url,
-              donate: Exercism::Routes.donate_url
+              confirm_params_return_url: Exercism::Routes.donate_url
```
Comment (ErikSchierboom) [inline app/helpers/react_components/donations/footer_form.rb]: I'm not entirely sure about the name. What is this link used for? We usually don't use a `_url` suffix

File: app/javascript/components/modals/student/FinishMentorDiscussionModal.tsx [hunk 14]
```diff
@@ -72,23 +72,36 @@ const Inner = ({
...
+          <Step.DonationStep
+            donation={donation}
+            links={links}
+            onSuccessfulDonation={(_, amount) => {
+              setDonatedAmount(amount)
+              send('SUCCESSFUL_DONATION')
+            }}
+          />
+        )
+      } else
```
Comment (ErikSchierboom) [inline app/javascript/components/modals/student/FinishMentorDiscussionModal.tsx]: You could lose the `else` and just directly `return` here

File: app/javascript/components/types.ts [hunk 15]
```diff
@@ -453,6 +453,26 @@ export type MentoredTrackExercise = {
   completedByMentor: boolean
 }
 
+export type MentoringSessionDonation = {
+  showDonationModal: boolean
+  request: {
+    endpoint: string
+    options: {
+      initialData: string
```
Comment (ErikSchierboom) [inline app/javascript/components/types.ts]: Is the initial data a `string`?

File: app/models/user.rb [hunk 16]
```diff
@@ -221,6 +221,10 @@ def formatted_reputation(*args)
     User::FormatReputation.(rep)
   end
 
+  def donated_in_last_35_days?
```
Comment (ErikSchierboom) [inline app/models/user.rb:225]:
```suggestion
  memoize
  def donated_in_last_35_days?
```

File: .github/labels.yml (hunk 11 above)
Comment (dem4ron) [reply .github/labels.yml]: This was merged from main at some point:
https://github.com/exercism/website/pull/6202/commits/28e1729898de403603625521195a01b501c07d55

File: CODE_OF_CONDUCT.md (hunk 12 above)
Comment (dem4ron) [reply CODE_OF_CONDUCT.md]: same as above: https://github.com/exercism/website/pull/6202/commits/28e1729898de403603625521195a01b501c07d55

File: app/helpers/react_components/donations/footer_form.rb (hunk 13 above)
Comment (dem4ron) [reply app/helpers/react_components/donations/footer_form.rb]: This goes into `stripe.confirmPayment.options.confirmParams.[returnUrl | return_url]`

File: app/javascript/components/types.ts (hunk 15 above)
Comment (dem4ron) [reply app/javascript/components/types.ts]: Updated, thanks!

File: app/helpers/react_components/student/mentoring_session.rb [hunk 17]
```diff
@@ -95,6 +110,14 @@ def iterations
 
         SerializeIterations.(solution.iterations, comment_counts:)
       end
+
+      def show_donation_modal
+        return false if current_user.insider?
+        return false if current_user.donated_in_last_35_days?
+
+        num_testimonials = current_user.provided_testimonials.count
+        num_testimonials.zero? || (num_testimonials % 3).zero?
```
Comment (ErikSchierboom) [inline app/helpers/react_components/student/mentoring_session.rb]: 0 % 3 = 0

```suggestion
        (num_testimonials % 3).zero?
```

File: app/css/ui-kit/filters.css (hunk 2 above)
Comment (dem4ron) [reply app/css/ui-kit/filters.css]: This was actually unused. It might have been used at some point in this PR, but upon doing a grep, I couldn't find any references to it. I've removed it.

File: .github/labels.yml (hunk 11 above)
Comment (iHiD) [reply .github/labels.yml]:
```suggestion
  
```

File: CODE_OF_CONDUCT.md (hunk 12 above)
Comment (iHiD) [reply CODE_OF_CONDUCT.md]:
```suggestion
This policy applies to the Exercism website, the Exercism GitHub organization, any other Exercism-related communication channels (e.g.  Slack, Twitter, email) and any other Exercism entity or event._
```

File: app/helpers/react_components/donations/footer_form.rb (hunk 13 above)
Comment (ErikSchierboom) [reply app/helpers/react_components/donations/footer_form.rb]: I would probably rename it slightly to something like `donate_return_url`.

Comment (iHiD) [reply app/helpers/react_components/donations/footer_form.rb]: I think I'm ok with the `confirm_params_return` bit as it's in the scope of the form, but agree it probably shouldn't have the `_url` prefix. Pragmatically I don't care if its hard to change it, but if it's just a couple of renames it would be better to be consistent.

Comment (dem4ron) [reply app/helpers/react_components/donations/footer_form.rb]: I agree, but the reason I chose this name is because this is exactly what the API accepts. so it is pretty specific.

Comment (ErikSchierboom) [reply app/helpers/react_components/donations/footer_form.rb]: Well, the reason I was suggesting `donate_return_url` (actually, `redirect_to_after_donation` or something like that might be better) is because it isn't coupled to Stripe. I feel that's an implementation detail 🤷
